File: 프로그래머스/lv2/최원준/다리를 지나는 트럭 .py
'''
1. 아이디어 :
    큐로 구현
2. 시간복잡도 :
    O( n )
3. 자료구조 :
    큐
'''

from collections import deque
def solution(bridge_length, weight, truck_weights):
    total = 0
    count = 0
    arrived = len(truck_weights)
    truck_weights = deque(truck_weights)
    queue = deque([0] * bridge_length)

    while arrived:
        if queue[0] != 0:
            arrived -= 1
        total -= queue.popleft()

        if truck_weights and total + truck_weights[0] <= weight:
            total += truck_weights[0]
            queue.append(truck_weights.popleft())
        else:
            queue.append(0)

        count += 1
    return count

File: 프로그래머스/lv2/지준혁/다리를 지나는 트럭.java
import java.util.*;

// 트럭 현재 무게, 다리를 건너는데 걸리는 시간
// 1. 다리 하중보다 낮다면 현재 하중이 낮다면 계속 트럭 전진시킴
// 2. 다리를 건넌 트럭이 있다면 현재 하중에 트럭 무게 차감
// 3. 마지막으로 건넌 트럭이 다리를 건너는 데 걸리는 시간이 모든 트럭이 지나는데 걸리는 시간임

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {

        Queue<int[]> q = new LinkedList<>();
        int i = 0;
        int time = 0;
        int curWeight = 0;
        while (i < truck_weights.length) {
            if (curWeight + truck_weights[i] <= weight) {
                q.add(new int[]{truck_weights[i], time + bridge_length + 1});
                curWeight += truck_weights[i];
                ++i;
            }
            ++time;
            if (q.peek()[1] == time + 1) {
                curWeight -= q.poll()[0];
            }
        }
        int ans = 0;
        while (!q.isEmpty()) {
            ans = q.poll()[1];
        }
        return ans;
    }
}
